modules/katello_sync_plan.py

```diff
@@ -84,154 +92,89 @@
-            if not self._module.check_mode:
-                product_ids_to_add = set(desired_product_ids) - set(current_product_ids)
-                if len(product_ids_to_add) > 0:
-                    sync_plan.add_products(data={'product_ids': list(product_ids_to_add)})
-                product_ids_to_remove = set(current_product_ids) - set(desired_product_ids)
-                if len(product_ids_to_remove) > 0:
-                    sync_plan.remove_products(data={'product_ids': list(product_ids_to_remove)})
-            updated = True
 
-        return updated
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'description',
+    'organization': 'organization',
+    'interval': 'interval',
+    'enabled': 'enabled',
+    'sync_date': 'sync_date',
+    'cron_expression': 'cron_expression',
+    'products': 'product',
+}
 
 
 def main():
-    module = ForemanAnsibleModule(
+    module = ForemanEntityAnsibleModule(
         argument_spec=dict(
             name=dict(required=True),
+            description=dict(),
             organization=dict(required=True),
-            interval=dict(required=True),
-            enabled=dict(required=True),
+            interval=dict(choices=['hourly', 'daily', 'weekly', 'custom cron'], required=True),
+            enabled=dict(type='bool', required=True),
             sync_date=dict(required=True),
-            products=dict(type='list', default=[]),
+            cron_expression=dict(),
+            products=dict(type='list'),
         ),
         supports_check_mode=True
     )
 
-    module_params = module.parse_params()
-    server_params = module.get_server_params()
-
-    name = module_params['name']
-    organization = module_params['organization']
-    interval = module_params['interval']
-    enabled = module_params['enabled']
-    sync_date = datetime.strptime(module_params['sync_date'], '%Y-%m-%d %H:%M:%S').replace(tzinfo=pytz.UTC)
-    products = module_params['products']
-
-    (server_url, username, password, verify_ssl) = server_params
-    server = ServerConfig(
-        url=server_url,
-        auth=(username, password),
-        verify=verify_ssl
-    )
-    ng = NailGun(server, entities, module)
-
-    # Lets make an connection to the server with username and password
-    try:
-        org = entities.Organization(server)
-        org.search()
-    except Exception as e:
-        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
-
-    try:
-        changed = ng.sync_plan(name, organization, interval=interval, enabled=enabled, sync_date=sync_date, products=products)
-        module.exit_json(changed=changed)
-    except Exception as e:
-        module.fail_json(msg=e)
+    (entity_dict, state) = module.parse_params()
+
+    if entity_dict['interval'] == 'custom cron':
+        if 'cron_expression' not in entity_dict:
+            module.fail_json(msg='"cron_expression" is missing.')
```

Comment:
You can do this with `required_if` instead https://github.com/theforeman/foreman-ansible-modules/blob/fa9bfb091523de1cfd713210e4402ea82aef7065/modules/katello_manifest.py#L172-L174